keps/sig-api-machinery/4008-crd-ratcheting/README.md

```diff
@@ -323,6 +335,146 @@ myOtherField: newly added field
+nested within the following will be not ratcheted:
+
+- not
+- oneOf
+
+Our intuition is that in the wild CRD authors do not make much use of them, and
+when they do the logic is complicated and references multiple fields. 
+
+For all these constructs, the conservative `DeepEqual` rule still applies 
+to the final output of the rule. 
+
+The following will be ratcheted with the conservative ratcheting rule, and
+their children will also be ratcheted.
+
+- allOf
+- anyOf
+
+
+### XListType and XMapKeys
+
+Errors thrown due to changing a list to a map-list with `x-list-type: map`or 
+changing its `x-kubernetes-map-keys` validations will not be ratched. 
+
+For example:
+```yaml
+- namespace: myNS
+  name: myObj
+- name: myObj
+```
+
+Changing `x-kubernetes-map-keys: [namespace] to `x-kubernetes-map-keys: [name, namespace]`
+would trigger a new `Duplicate key` error not previously seen for this object.
+
+These fields are not ratcheted because changing map keys or list type to a map alters 
+the structure of the schema, and old values cannot be correlated to new. 
+
+Additionally Server-Side-Apply does not tolerate updates that violate the map 
+keys constraint: these properties are unsafe to change without bumping CRD 
+version.
+
+### Atomic Lists and Maps
+
+Following Kubernetes convention that fields annotated `x-kubernetes-list-type`
+or `x-kubernetes-map-type` as `atomic` change as a unit, the ratcheting rules
+do not apply to values nested inside atomic lists and maps. But a DeepEqual 
+comparison on the entire object can still ratchet errors for subfields.
+
+### Weaker Ratcheting Rule for Value Validations
+
+Above we propose a conservative catch-all ratcheting rule that should provide
+a minimum level of ratcheting support for any validation rule in the schema.
+
+`DeepEqual(OldObject, NewObject)` requirement is very strong. But for some 
+validation rules it is too strict. Consider the `required` error. This
+validation is applied to the object owning the property. Thus, if a new field
+becomes required the object cannot have other fields added onto it or changed
+without also adding the newly required field.
+
+Kubernetes has precedent for using a weaker form of racheting when validation is
+strengthened:
+
+```go
+if !Validation(old) {
+  // Ignore errors caused by !Validation(new)
+}
+```
+
+An example of this is the structural schema requirement added when Kubernetes
+went GA ([source](https://github.com/kubernetes/kubernetes/blob/12dc19d46fb97cbbfeb1e12b8a10ff7ae73d9515/staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go#L1583-L1590)): 
+```go
+// requireStructuralSchema returns true if schemas specified must be structural
+func requireStructuralSchema(oldCRDSpec *apiextensions.CustomResourceDefinitionSpec) bool {
+	if oldCRDSpec != nil && specHasNonStructuralSchema(oldCRDSpec) {
+		// don't tighten validation on existing persisted data
+		return false
+	}
```

Comment:
I read this as .. "if I had a non-structural schema, I can update it to another non-structural schema.". Something that is not captured here is whether the change makes it closer or farther from being non-structural. It's fine because being non-structural is an all or nothing thing, so there's not a practical "being further or closer to being structural", or at least we chose not to care.

Now I think you're saying: "if I have an object that misses a field that is required, I can update it to another object that misses required fields", you can actually be further from the goal by introducing new required fields (that wouldn't have passed before). There's definitely a practical meaning of being "farther or closer". I guess it's easy to decide that if a field is required it shouldn't be removable, but that's unclear from the text. I'm not even sure I'm thinking about this the right way.

Reply:
After discussion with antoine online, we agreed a goal of this KEP should be to only allow values that we know for certain pass either the old schema or the new schema. We found that most constraints are not possible to have a weaker ratcheting condition with this goal in mind, and the ones that do are also not possible to ratchet if CEL is used on a parent.